Baikal/SceneGraph/texture.cpp

```diff
@@ -62,29 +58,127 @@ namespace Baikal
                 avg += RadeonRays::float3(r, g, b);
             }
 
-            avg *= (1.f / num_elements);
             break;
         }
         default:
-            break;
+            throw std::runtime_error("Texture::ComputeAverageValue(): unsupported format");
         }
 
+        avg *= (1.f / num_elements);
+
         return avg;
     }
 
     namespace {
         struct TextureConcrete : public Texture {
             TextureConcrete() = default;
-            TextureConcrete(char* data, RadeonRays::int3 size, Format format) :
-                Texture(data, size, format) {}
+            TextureConcrete(
+                char* data,
+                const std::vector<RadeonRays::int3> &levels,
+                Format format) :
+                Texture(data, levels, format) {}
+
+            TextureConcrete(char* data, RadeonRays::int3 size, Format format, bool generate_mipmap) :
+                Texture(data, size, format, generate_mipmap) {}
         };
     }
 
     Texture::Ptr Texture::Create() {
         return std::make_shared<TextureConcrete>();
     }
 
-    Texture::Ptr Texture::Create(char* data, RadeonRays::int3 size, Format format) {
-        return std::make_shared<TextureConcrete>(data, size, format);
+    Texture::Ptr Texture::Create(char* data, const std::vector<RadeonRays::int3> &levels, Format format)
+    {
+        return std::make_shared<TextureConcrete>(data, levels, format);
+    }
+
+    Texture::Ptr Texture::Create(char* data, RadeonRays::int3 size, Format format, bool generate_mipmap)
+    {
+        return std::make_shared<TextureConcrete>(data, size, format, generate_mipmap);
+    }
+
+    Texture::Texture(
+        char* data,
+        std::vector<RadeonRays::int3> const& mip_sizes,
+        Format format)
+        : m_data(data)
+        , m_format(format)
+        , m_needs_mip_generation(false)
+    {
+        if (mip_sizes.empty())
+        {
+            throw std::logic_error("Texture::Texture(...): mip_sizes is empty");
+        }
+
+        m_mip_sizes = mip_sizes;
+
+        std::transform(m_mip_sizes.begin(), m_mip_sizes.end(), m_mip_sizes.begin(),
```

Comment:
What are you doing here?

Reply:
there are only 2 places where user can set size. So please check size at set-time

Reply:
Ok, I've done this